pkg/credentialprovider/external.go

```diff
@@ -0,0 +1,224 @@
+	codecs = serializer.NewCodecFactory(scheme)
+}
+
+type externalProviderKeyring struct {
+	// map of registry matcher to provider (each provider may have multiple)
+	providers map[string]ExternalCredentialProvider
+	// reverse sorted index of registries
+	index []string
+}
+
+func NewExternalProviderKeyring(configPath string) (DockerKeyring, error) {
+	keyring := &externalProviderKeyring{
+		providers: make(map[string]ExternalCredentialProvider),
+		index:     make([]string, 0),
+	}
+
+	config, err := readExternalProviderConfig(configPath)
+	if err != nil {
+		return keyring, err
+	}
+
+	for _, p := range config.Providers {
+		provider := externalCredentialProvider{
+			command: p.Exec.Command,
+			args:    p.Exec.Args,
+			env:     convertEnvs(p.Exec.Env),
+		}
+		for _, m := range p.ImageMatchers {
+
+			value := m
+			if !strings.HasPrefix(value, "https://") && !strings.HasPrefix(value, "http://") {
+				value = "https://" + value
+			}
+			parsed, err := url.Parse(value)
+			if err != nil {
+				klog.Errorf("Entry %q in dockercfg invalid (%v), ignoring", m, err)
+				continue
+			}
+
+			// The docker client allows exact matches:
+			//    foo.bar.com/namespace
+			// Or hostname matches:
+			//    foo.bar.com
+			// It also considers /v2/  and /v1/ equivalent to the hostname
+			// See ResolveAuthConfig in docker/registry/auth.go.
+			effectivePath := parsed.Path
+			if strings.HasPrefix(effectivePath, "/v2/") || strings.HasPrefix(effectivePath, "/v1/") {
+				effectivePath = effectivePath[3:]
+			}
+			var key string
+			if (len(effectivePath) > 0) && (effectivePath != "/") {
+				key = parsed.Host + effectivePath
+			} else {
+				key = parsed.Host
+			}
+
+			keyring.providers[key] = &provider
+			keyring.index = append(keyring.index, key)
+		}
+	}
+
+	eliminateDupes := sets.NewString(keyring.index...)
+	keyring.index = eliminateDupes.List()
+
+	// Update the index used to identify which credentials to use for a given
+	// image. The index is reverse-sorted so more specific paths are matched
+	// first. For example, if for the given image "quay.io/coreos/etcd",
+	// credentials for "quay.io/coreos" should match before "quay.io".
+	sort.Sort(sort.Reverse(sort.StringSlice(keyring.index)))
+
+	// TODO remove
+	klog.V(2).Infof("Contents of keyring index %v", keyring.index)
+	klog.V(2).Infof("Contents of keyring providers %v", keyring.providers)
+
+	return keyring, nil
+}
+
+func (e *externalProviderKeyring) Lookup(image string) ([]AuthConfig, bool) {
+	var provider ExternalCredentialProvider
+	klog.V(2).Infof("Looking up %s with external provider.", image)
+
+	// range over the index as iterating over a map does not provide a predictable ordering
+	for _, matcher := range e.index {
+		// both k and image are schemeless URLs because even though schemes are allowed
+		// in the credential configurations, we remove them in Add.
+		if matched, _ := urlsMatchStr(matcher, image); matched {
+			provider = e.providers[matcher]
```

Comment:
Are we deliberately picking the last matched provider?

Reply:
No, good catch